File: LeetCode/Java/src/main/java/org/redquark/tutorials/leetcode/LongestSubstringWithoutRepeatingCharacters.java
package org.redquark.tutorials.leetcode;

import java.util.HashSet;
import java.util.Set;

/**
 * @author Anirudh Sharma
 * <p>
 * Given a string s, find the length of the longest substring without repeating characters.
 */
public class LongestSubstringWithoutRepeatingCharacters {

    /**
     * @param s - given input string
     * @return - length of longest substring without repeating characters
     */
    private static int lengthOfLongestSubstring(String s) {
        // Base condition
        if (s == null || s.equals("")) {
            return 0;
        }
        // Starting window index
        int start = 0;
        // Ending window index
        int end = 0;
        // Maximum length of substring
        int maxLength = 0;
        // This set will store the unique characters
        Set<Character> uniqueCharacters = new HashSet<>();
        while (end < s.length()) {
            if (uniqueCharacters.add(s.charAt(end))) {
                end++;
                maxLength = Math.max(maxLength, uniqueCharacters.size());
            } else {
                uniqueCharacters.remove(s.charAt(start));
                start++;
            }
        }
        return maxLength;
    }

    public static void main(String[] args) {
        String s = "abcabcbb";
        System.out.println(lengthOfLongestSubstring(s));

        s = "bbbbb";
        System.out.println(lengthOfLongestSubstring(s));

        s = "pwwkew";
        System.out.println(lengthOfLongestSubstring(s));

        s = "";
        System.out.println(lengthOfLongestSubstring(s));

        s = "aab";
        System.out.println(lengthOfLongestSubstring(s));
    }
}

File: LeetCode/Kotlin/src/main/kotlin/org/redquark/tutorials/leetcode/LongestSubstringWithoutRepeatingCharacters.kt
package org.redquark.tutorials.leetcode


private fun lengthOfLongestSubstring(s: String): Int {
    // Base condition
    if (s == "") {
        return 0
    }
    // Starting window index
    var start = 0
    // Ending window index
    var end = 0
    // Maximum length of substring
    var maxLength = 0
    // This set will store the unique characters
    val uniqueCharacters: MutableSet<Char> = HashSet()
    while (end < s.length) {
        if (uniqueCharacters.add(s[end])) {
            end++
            maxLength = maxLength.coerceAtLeast(uniqueCharacters.size)
        } else {
            uniqueCharacters.remove(s[start])
            start++
        }
    }
    return maxLength
}

fun main() {
    var s = "abcabcbb"
    println(lengthOfLongestSubstring(s))

    s = "bbbbb"
    println(lengthOfLongestSubstring(s))

    s = "pwwkew"
    println(lengthOfLongestSubstring(s))

    s = ""
    println(lengthOfLongestSubstring(s))

    s = "aab"
    println(lengthOfLongestSubstring(s))
}